File: StoryTime2.0/Models/AnthropicClient.swift
import Foundation

// MARK: - Anthropic client
//
// Minimal bring-your-own-API-key wrapper around the Anthropic Messages
// API. Used to generate a personal CYOA story by sending the model a
// strict instruction to return our CatalogStory JSON schema.
//
// No data leaves the device except to api.anthropic.com when the user
// taps "Generate."

enum AnthropicError: Error, LocalizedError {
    case missingKey
    case http(Int, String)
    case decode(String)
    case noContent
    case invalidGraph([StoryGraphValidator.Issue])

    var errorDescription: String? {
        switch self {
        case .missingKey:   return "Add an Anthropic API key in Settings first."
        case .http(let s, let body): return "API error (\(s)): \(body.prefix(200))"
        case .decode(let why): return "Couldn't read the response: \(why)"
        case .noContent:    return "The model returned no usable content."
        case .invalidGraph(let issues):
            return "The generated story had broken scenes (\(issues.map(\.description).joined(separator: "; ").prefix(200))). Please try again."
        }
    }
}

struct AnthropicClient {
    /// Default model. Latest Claude family that's good at structured generation.
    var model: String = "claude-sonnet-4-5"
    var apiKey: String

    /// Generate a story and validate its node graph. If the first attempt
    /// produces a broken graph (the most common model failure), retry once
    /// with the validation errors fed back; if that also fails, surface
    /// the issues to the caller.
    func generateStory(title: String, premise: String, kind: StoryKind, genre: StoryGenre) async throws -> CatalogStory {
        let first = try await generateStoryOnce(title: title, premise: premise,
                                                kind: kind, genre: genre)
        let issues = StoryGraphValidator.validate(first)
        if issues.isEmpty { return first }

        let feedback = "Your previous attempt had these structural errors: "
            + issues.map(\.description).joined(separator: "; ")
            + ". Regenerate the full JSON with all of them fixed."
        let second = try await generateStoryOnce(title: title, premise: premise,
                                                 kind: kind, genre: genre,
                                                 extraInstruction: feedback)
        let secondIssues = StoryGraphValidator.validate(second)
        guard secondIssues.isEmpty else {
            throw AnthropicError.invalidGraph(secondIssues)
        }
        return second
    }

    private func generateStoryOnce(title: String, premise: String, kind: StoryKind,
                                   genre: StoryGenre, extraInstruction: String? = nil) async throws -> CatalogStory {
        guard !apiKey.trimmingCharacters(in: .whitespaces).isEmpty else {
            throw AnthropicError.missingKey
        }

        let system = """
        You are a CYOA story author. Reply with ONE JSON object only, no \
        prose before or after. It must conform to this schema (Swift Codable):

        {
          "id": "user-<slug>-####",
          "title": "<the original work's title>",
          "sourceTitle": "<the original work's title>",
          "kind": "movie" | "show" | "book",
          "synopsis": "<one to two sentences, second-person>",
          "releaseYear": <int or null>,
          "addedAt": "<ISO 8601 timestamp>",
          "genre": "Drama" | "Comedy" | "Thriller" | "Sci-Fi" | "Horror" | "Fantasy" | "Action",
          "tags": [<short strings>],
          "rating": null,
          "loved": false,
          "nextStoryId": null,
          "startNodeId": "s1",
          "nodes": [ <see below> ]
        }

        Nodes form a directed graph. Use 12–14 decision nodes named s1..sN \
        and 3 ending nodes named end_a, end_b, end_c. For decision nodes, \
        "choices" is an array of 2 objects with text/consequence/nextNodeId \
        and "isEnding" is false. Endings: "choices": [], "isEnding": true, \
        "endingTitle" set. Make the prose vivid, second-person, with \
        "what would you have done in <Title>?" framing. Every nextNodeId \
        must exist. The last decision node's choices point to the endings.
        """

        var user = """
        Title: \(title)
        Kind: \(kind.rawValue)
        Genre: \(genre.rawValue)
        Premise / framing: \(premise.isEmpty ? "Place the reader in the world of the original." : premise)

        Generate the JSON now.
        """
        if let extraInstruction {
            user += "\n\n" + extraInstruction
        }

        let body: [String: Any] = [
            "model": model,
            "max_tokens": 8192,
            "system": system,
            "messages": [
                ["role": "user", "content": user]
            ]
        ]

        var request = URLRequest(url: URL(string: "https://api.anthropic.com/v1/messages")!)
        request.httpMethod = "POST"
        request.setValue(apiKey, forHTTPHeaderField: "x-api-key")
        request.setValue("2023-06-01", forHTTPHeaderField: "anthropic-version")
        request.setValue("application/json", forHTTPHeaderField: "content-type")
        request.httpBody = try JSONSerialization.data(withJSONObject: body)

        let (data, response) = try await URLSession.shared.data(for: request)
        if let http = response as? HTTPURLResponse, !(200..<300).contains(http.statusCode) {
            let s = String(data: data, encoding: .utf8) ?? ""
            throw AnthropicError.http(http.statusCode, s)
        }

        // Anthropic returns { content: [{ type: "text", text: "..." }] }
        guard let json = try JSONSerialization.jsonObject(with: data) as? [String: Any],
              let parts = json["content"] as? [[String: Any]] else {
            throw AnthropicError.decode("unexpected response shape")
        }
        let combined = parts.compactMap { $0["text"] as? String }.joined()
        guard !combined.isEmpty else { throw AnthropicError.noContent }

        // Carve out the first JSON object from the response.
        guard let storyJSONString = Self.extractFirstJSONObject(from: combined),
              let storyData = storyJSONString.data(using: .utf8) else {
            throw AnthropicError.decode("no JSON object found in response")
        }
        do {
            return try CatalogJSON.decoder.decode(CatalogStory.self, from: storyData)
        } catch {
            throw AnthropicError.decode(String(describing: error))
        }
    }

    /// Pull the first balanced `{...}` block out of a string. Tolerant of
    /// minor preamble or trailing prose around the JSON.
    static func extractFirstJSONObject(from text: String) -> String? {
        guard let start = text.firstIndex(of: "{") else { return nil }
        var depth = 0
        var inString = false
        var escape = false
        var i = start
        while i < text.endIndex {
            let c = text[i]
            if escape { escape = false; i = text.index(after: i); continue }
            if c == "\\" && inString { escape = true; i = text.index(after: i); continue }
            if c == "\"" { inString.toggle() }
            else if !inString {
                if c == "{" { depth += 1 }
                if c == "}" {
                    depth -= 1
                    if depth == 0 {
                        return String(text[start...i])
                    }
                }
            }
            i = text.index(after: i)
        }
        return nil
    }
}

File: StoryTimeAndroid/app/src/main/java/com/storytime/android/ai/AnthropicClient.kt
package com.storytime.android.ai

import com.storytime.android.model.CatalogStory
import com.storytime.android.model.StoryKind
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.buildJsonArray
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.contentOrNull
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import kotlinx.serialization.json.put
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import java.util.concurrent.TimeUnit

class AnthropicError(message: String) : Exception(message)

class AnthropicClient(
    private val apiKey: String,
    private val model: String = "claude-sonnet-4-5",
) {
    private val client = OkHttpClient.Builder()
        .callTimeout(60, TimeUnit.SECONDS)
        .readTimeout(60, TimeUnit.SECONDS)
        .build()

    private val json = Json { ignoreUnknownKeys = true; coerceInputValues = true }

    suspend fun generateStory(title: String, premise: String, kind: StoryKind, genre: String): CatalogStory =
        withContext(Dispatchers.IO) {
            if (apiKey.isBlank()) throw AnthropicError("Add an Anthropic API key in Settings first.")

            val system = SYSTEM_PROMPT
            val user = "Title: $title\nKind: ${kind.name.lowercase()}\nGenre: $genre\n" +
                "Premise / framing: ${premise.ifEmpty { "Place the reader in the world of the original." }}\n\nGenerate the JSON now."

            val body = buildJsonObject {
                put("model", model)
                put("max_tokens", 8192)
                put("system", system)
                put("messages", buildJsonArray {
                    add(buildJsonObject {
                        put("role", "user")
                        put("content", user)
                    })
                })
            }

            val req = Request.Builder()
                .url("https://api.anthropic.com/v1/messages")
                .addHeader("x-api-key", apiKey)
                .addHeader("anthropic-version", "2023-06-01")
                .addHeader("content-type", "application/json")
                .post(body.toString().toRequestBody("application/json".toMediaType()))
                .build()

            client.newCall(req).execute().use { resp ->
                val raw = resp.body?.string().orEmpty()
                if (!resp.isSuccessful) throw AnthropicError("API error (${resp.code}): ${raw.take(200)}")
                val parsed = json.parseToJsonElement(raw).jsonObject
                val combined = (parsed["content"] as? JsonArray)
                    ?.mapNotNull { (it as? JsonObject)?.get("text")?.jsonPrimitive?.contentOrNull }
                    ?.joinToString("")
                    .orEmpty()
                if (combined.isBlank()) throw AnthropicError("The model returned no usable content.")
                val storyJson = extractFirstJsonObject(combined)
                    ?: throw AnthropicError("No JSON object found in response.")
                runCatching { json.decodeFromString(CatalogStory.serializer(), storyJson) }
                    .getOrElse { throw AnthropicError("Couldn't decode story: ${it.message}") }
            }
        }

    companion object {
        private val SYSTEM_PROMPT = """
            You are a CYOA story author. Reply with ONE JSON object only, no prose before or after.
            Schema:
            {
              "id": "user-<slug>-####",
              "title": "<the original work's title>",
              "sourceTitle": "<same>",
              "kind": "movie" | "show" | "book",
              "synopsis": "<one to two sentences, second-person>",
              "releaseYear": <int or null>,
              "addedAt": "<ISO 8601 timestamp>",
              "genre": "Drama" | "Comedy" | "Thriller" | "Sci-Fi" | "Horror" | "Fantasy" | "Action",
              "tags": [<short strings>],
              "rating": null,
              "loved": false,
              "nextStoryId": null,
              "startNodeId": "s1",
              "nodes": [ ... ]
            }
            Nodes form a directed graph: 12–14 decision nodes named s1..sN and 3 endings end_a/end_b/end_c.
            Decision nodes: "isEnding": false, "choices": [2 objects with text/consequence/nextNodeId].
            Endings: "isEnding": true, "choices": [], "endingTitle" set. Every nextNodeId must exist.
            Vivid second-person prose, "what would you have done in <Title>?" framing.
        """.trimIndent()

        fun extractFirstJsonObject(text: String): String? {
            val start = text.indexOf('{')
            if (start < 0) return null
            var depth = 0
            var inString = false
            var escape = false
            for (i in start until text.length) {
                val c = text[i]
                if (escape) { escape = false; continue }
                if (c == '\\' && inString) { escape = true; continue }
                if (c == '"') { inString = !inString; continue }
                if (inString) continue
                if (c == '{') depth++
                if (c == '}') {
                    depth--
                    if (depth == 0) return text.substring(start, i + 1)
                }
            }
            return null
        }
    }
}
